README.md
## Z-Data

[![Hackage](https://img.shields.io/hackage/v/Z-Data.svg?style=flat)](https://hackage.haskell.org/package/Z-Data)
[![Linux Build Status](https://github.com/ZHaskell/z-data/workflows/ubuntu-ci/badge.svg)](https://github.com/ZHaskell/z-data/actions)
[![macOS Build Status](https://github.com/ZHaskell/z-data/workflows/osx-ci/badge.svg)](https://github.com/ZHaskell/z-data/actions)
[![Windows Build Status](https://github.com/ZHaskell/z-data/workflows/win-ci/badge.svg)](https://github.com/ZHaskell/z-data/actions)
[![ARM Build Status](https://img.shields.io/drone/build/ZHaskell/z-data?label=arm-ci)](https://cloud.drone.io/ZHaskell/z-data)
[![Gitter chat](https://badges.gitter.im/gitterHQ/gitter.svg)](https://gitter.im/Z-Haskell/community)
<a href="https://opencollective.com/zhaskell/donate" target="_blank">
  <img src="https://opencollective.com/zhaskell/donate/button@2x.png?color=blue" width=128 />
</a>

This package is part of [ZHaskell](https://z.haskell.world) project, providing basic data structures and functions:

* Array, vector(array slice), sorting, searching
* Text based UTF-8, basic unicode manipulating, regex
* FFI utilties
* Fast parsing and building monad
* JSON encoding and decoding

## Requirements

* A working haskell compiler system, GHC(>=8.6), cabal-install(>=2.4), here're some options:
    * Mac users can get them via [homebew](//brew.sh/): `brew install ghc cabal-install`.
    * Windows users can get them via [chocolatey](//chocolatey.org): `choco install ghc cabal`.
    * Ubuntu users are recommended to use this [ppa](//launchpad.net/~hvr/+archive/ubuntu/ghc).
* A working C/C++ compiler support C++11, here're some options:
    * Mac users can use the `clang` comes with the [XCode](https://developer.apple.com/xcode/) or [XCode Command Line Tools](https://developer.apple.com/downloads):
        * You can install XCode from app store, or XCode Command Line Tools with `sudo xcode-select --install`.
        * If you came across compiling issues like [this](https://stackoverflow.com/questions/58628377/catalina-c-using-cmath-headers-yield-error-no-member-named-signbit-in-th), consider removing redundant SDKs, e.g. `sudo rm -rf /Library/Developer/CommandLineTools/SDKs`
    * Windows users can use the mingw's one comes with GHC, you can use it by adding `your_path_to_ghc\mingw\bin` to your `PATH`.
    * Ubuntu users can install `gcc/g++` by running `sudo apt install build-essential`.
* Tests need [hspec-discover](https://hackage.haskell.org/package/hspec-discover).

## Example usage

```haskell
> import qualified Z.Data.Vector as V
> import qualified Z.Data.Array as A
>
> -- convert from list
> let v = V.pack [1..10] :: V.PrimVector Int  
> -- vector combinators works on arrays as well
> let a = V.pack [1..10] :: A.Array Int   
> -- slicing vector(slice) is O(1)
> V.take 3 v                              
[1,2,3]
-- slicing array is not O(1)
> V.drop 3 a                              
fromListN 7 [4,5,6,7,8,9,10]
>
> V.intersperse 10 v
[1,10,2,10,3,10,4,10,5,10,6,10,7,10,8,10,9,10,10]
>
> V.mergeSort (V.intersperse 10 v) 
[1,2,3,4,5,6,7,8,9,10,10,10,10,10,10,10,10,10,10]
> -- Generic KMP search on vectors
> V.indices (V.singleton 10) (V.intersperse 10 v) True   
[1,3,5,7,9,11,13,15,17,18]
>
> -- quoter for writing numeric vector literals
> :set -XQuasiQuotes 
> :t [V.vecWord|1,2,3,4,5,4,3,2,1|]                     
[V.vecWord|1,2,3,4,5,4,3,2,1|] :: V.PrimVector Word
>
> import qualified Z.Data.Builder as B
> import qualified Z.Data.Text as T
> :set -XOverloadedStrings 
>
> -- Builders can be used with OverloadedStrings
> B.build $ "builders: " >> B.hex (3 :: Word16) >> B.comma >> B.double 1.2345678
[98,117,105,108,100,101,114,115,58,32,48,48,48,51,44,49,46,50,51,52,53,54,55,56]
> 
> B.buildText $ "builders: " >> B.hex (3 :: Word16) >> B.comma >> B.double 1.2345678
"builders: 0003,1.2345678"
>
> import qualified Z.Data.JSON as JSON
> import GHC.Generics
> 
> JSON.parseValue "[1,2,3,4,5]"
([],Right (Array [Number 1.0,Number 2.0,Number 3.0,Number 4.0,Number 5.0]))
>
> -- JSON module support deriving through Generic
> :set -XDeriveAnyClass -XDeriveGeneric
> data Foo = Foo {foo :: Double} deriving (JSON.JSON, Generic)
> JSON.toValue (Foo 0.01)
Object [("foo",Number 1.0e-2)]
> JSON.encodeText (Foo 0.01)
"{\"foo\":1.0e-2}"
```

## Dev guide

```bash
# get code
git clone --recursive git@github.com:ZHaskell/z-data.git 
cd z-data
# build
cabal build
# test
cabal test --test-show-details=direct
# install 
cabal install
# generate document
cabal haddock
```


ChangeLog.md
# Revision history for Z-Data

## 2.0.0.0  -- 2021-12-08

* Work only with GHC >= 9.2, use sized primitive types, new integer types.
* Remove dependencies on `ghc-prim`, `integer-gmp`, use modules exported by `base` instead. 
* Change `emptyArr` to a `Arr` class method.

## 1.1.0.0  -- 2021-07-15

* Fix building issues on ARM platform.
* Add `UUID` builders and parsers(both textual binary).
* Add more `PrimUnlifed` instances to `Z.Data.Array.UnliftedArray`.
* Add `doubleMutableArr` to `Z.Data.Array`, useful in some buffer building logic.
* Add `shuffle` and `permutations` to `Z.Data.Vector` and `Z.Data.Text`.
* Add `prettyJSON'` to `Z.Data.JSON` with custom indentation.
* Change `CBytes` 's JSON instance to write `__base64` field(instead of `base64` field) when not UTF8 encoded.
* Add missing type alias `UnliftedIORef` for `UnliftedRef RealWorld`.

## 1.0.0.1  -- 2021-07-08

* Fix a regression in `match` parsing combinator where matched chunk is returned instead of precise matched input.

## 1.0.0.0  -- 2021-07-05

* Clean up various `RULES` and `INLINE` pragmas, improve building time a little.
* Simplify `Z.Data.PrimRef` to use `PrimMonad`.
* Add `encodeXXX/encodeXXXLE/encodeXXXBE`(where `XXX` is a primitive type) to `Z.Data.Builder`.
* Add `check-array-bound` build flag to enable bound check in `Z.Data.Array` module, `Z.Data.Array.Checked` is removed.
* Add `concatR` to `Z.Data.Vector` and `Z.Data.Text`, which is useful to concat the result of an accumulator style recursive function.
* Improve date builder and parser by introducing faster common case path.  

## 0.9.0.0  -- 2021-07-01

* Add `decodeXXX/deocodeXXXLE/decodeXXXBE`(where `XXX` is a primitive type) to `Z.Data.Parser`.
* Rename `replicateMVec/traveseVec/traveseVec_` tp `replicateM/travese/travese_`, fix related `PrimMonad` rules not firing issue.
* Add a faster `sciToDouble` based on https://github.com/lemire/fast_double_parser, improve `double/double'` parser.

## 0.8.8.0  -- 2021-06-13

* Add `withCPtrForever` and `addCPtrDep` to `Z.Foreign.CPtr` module.
* Add `replicateMVec` to `Z.Data.Vector`.
* Fix AVX512 text validate header building with C11.

## 0.8.7.0  -- 2021-06-12

* Add `use-avx2` and `use-avx512` flags(default to False) to enable avx2 or avx512 base64 codec, UTF8 validator, etc.

## 0.8.6.1  -- 2021-06-09

* Change SIMD base64 codec to https://github.com/lemire/fastbase64.
* Add SIMD enabled `count` from bytestring.
* Switch `no-avx` on by default to improve compiled binary's portablity.
* Add `collate`, courtesy by jgm: https://github.com/jgm/unicode-collation.
* Add `hexDecodeWS/hexDecodeWS'` to `Z.Data.Vector.Hex`.

## 0.8.5.0  -- 2021-06-04

* Fix a bug in `peekMaybe` across buffer boundary.
* Clean up base64 AVX/SSE CPP flags.

## 0.8.4.0  -- 2021-06-03

* `arrVec` from `Z.Data.Vector` now could convert between vector(array slice) and array back and forth.
* Fix a CPP bug in `no-avx` build flag.

## 0.8.3.0  -- 2021-06-02

* Fix numeric literals' `Print` instances when precedence > 6 and  value < 0.
* Add `no-avx` build flag to run tests on VMs without AVX support.

## 0.8.2.0  -- 2021-05-10

* Add `withCPtrs`, `withCPtrsUnsafe` to `Z.Foreign.CPtr`.
* Add `displayWidthChar` to `Z.Data.Text`.

## 0.8.1.0  -- 2021-04-23

* Change `ParseChunks` and `parseChunks` to based on `parseChunk`.
* Remove `parseValueChunks`, add `decodeChunk` to `Z.Data.JSON`.

## 0.8.0.0  -- 2021-04-20

* Change `Bytes` 's `Hashable` instance to use FNV-1a (proposed by @zypeh).

## 0.7.4.0  -- 2021-04-06

* Add `emptyArr` to `Z.Data.Array`. 
* Add `newCPtr'` to accommodate common cases.
* Use `memchr` in `split/splitWith/lines` if possible.

## 0.7.3.0  -- 2021-03-30

* Add more helpers to debug `Parser`: `currentChunk`, `failWithInput`, `unsafeLiftIO`.
* Parser now is an instance of `PrimMonad`, which can perform limited effects, e.g. array operations.
* Make some parsers' error message more helpful.

## 0.7.2.0  -- 2021-03-22

* Add `fromMutablePrimArray` for constructing CBytes.
* Add `toBytes'/toBuilder'`(convert with the \NUL terminator) to `Z.Data.CBytes`.

## 0.7.1.0  -- 2021-03-03

* Add `CPtr` type, a more lightweight foreign pointer.
* Add `toLower/toUpper/toLowerLatin/toUpperLatin` to `Z.Data.ASCII`.

## 0.7.0.0  -- 2021-03-03

* Add more patterns to `Z.Data.ASCII`.
* Fix a bug in `Z.Data.Vector.Search` where needle is a singleton.
* Optimize `mergeDupAdjacent` family funtions in `Z.Data.Vector.Sort` to not copy vector in case of no duplication.
* Add `scientific'` to `Z.Data.Builder`, Add more `Print` instances.
* Bump deps to support GHC 9.0.

## 0.6.1.0  -- 2021-02-04

* Add `key` and `nth` lens to `Z.Data.JSON.Value` for manipulating nested value more easily.
* Port patch from bytestring #301 #310 #315, Improve `stime`, `sconcat`, `intersperse`.
* Add JSON pretty printer `prettyJSON/prettyValue` to `Z.Data.JSON`.
* Move many instances from `Z.Data.JSON.Base` to `Z.Data.JSON` to reduce the chance of heap overflow when compile.
* Add `modifyIndex/insertIndex/deleteIndex` to array and vector, rewrite `FlatMap/FlatSet/FlatInMap/FlatIntSet' to use them.

Z/Data/JSON.hs
{-# OPTIONS_GHC -fno-warn-orphans #-}

{-|
Module      : Z.Data.JSON
Description : Fast JSON serialization/deserialization
Copyright   : (c) Dong Han, 2020
License     : BSD
Maintainer  : winterland1989@gmail.com
Stability   : experimental
Portability : non-portable

Types and functions for working efficiently with JSON data, the design is quite similar to @aeson@ or @json@:

  * Encode to bytes can be done directly via 'encodeJSON'.
  * Decode are split in two step, first we parse JSON doc into 'Value', then convert to haskell data via 'fromValue'.
  * 'ToValue' are provided so that other doc formats can be easily supported, such as 'YAML'.

Note this module also provides many (orphan)instances to reduce the compilation stress of a gaint 'Z.Data.JSON.Base' module.

-}
module Z.Data.JSON
  ( -- * How to use this module
    -- $use
    -- ** Custom settings
    -- $custom-settings
    -- ** Write instances manually
    -- $manually-instance

    -- * JSON Class
    JSON(..), Value(..), defaultSettings, Settings(..)
  , snakeCase, trainCase
    -- * Encode & Decode
  , DecodeError
  , decode, decode', decodeText, decodeText'
  , ParseChunks, decodeChunk, decodeChunks
  , encode, encodeChunks, encodeText
  , prettyJSON, prettyValue, prettyJSON', prettyValue'
    -- * parse into JSON Value
  , parseValue, parseValue'
  -- * Generic functions
  , gToValue, gFromValue, gEncodeJSON
  -- * Convert 'Value' to Haskell data
  , convertValue, Converter(..), fail', (<?>), prependContext
  , PathElement(..), ConvertError(..)
  , typeMismatch, fromNull, withBool, withScientific, withBoundedScientific, withRealFloat
  , withBoundedIntegral, withText, withArray, withKeyValues, withFlatMap, withFlatMapR
  , withHashMap, withHashMapR, withEmbeddedJSON
  , (.:), (.:?), (.:!), convertField, convertFieldMaybe, convertFieldMaybe'
  -- * Helper for manually writing instance.
  , (.=), object, (.!), object', KVItem
  ) where

import           Data.Char
import           Data.Functor.Compose
import           Data.Functor.Const
import           Data.Functor.Identity
import           Data.Functor.Product
import           Data.Functor.Sum
import qualified Data.Monoid                    as Monoid
import           Data.Proxy                     (Proxy (..))
import           Data.Scientific                (toBoundedInteger)
import qualified Data.Semigroup                 as Semigroup
import           Data.Tagged                    (Tagged (..))
import           Data.Time                      (Day, DiffTime, LocalTime, NominalDiffTime, TimeOfDay, UTCTime, ZonedTime)
import           Data.Time.Calendar             (CalendarDiffDays (..), DayOfWeek (..))
import           Data.Time.LocalTime            (CalendarDiffTime (..))
import           Data.Time.Clock.System         (SystemTime (..))
import           Data.Version                   (Version(versionBranch), makeVersion)
import           Foreign.C.Types
import           System.Exit                    (ExitCode(..))
import qualified Z.Data.Builder                 as B
import           Z.Data.JSON.Base
import qualified Z.Data.Parser                  as P
import qualified Z.Data.Text                    as T
import           Data.UUID.Types                (UUID)

-- $use
--
-- This module is intended to be used qualified, e.g.
--
-- > import qualified Z.Data.JSON as JSON
-- > import           Z.Data.JSON ((.:), JSON(..))
--
-- The easiest way to use the library is to define target data type, deriving
-- 'GHC.Generics.Generic' and 'JSON' instances, which provides:
--
--   * 'fromValue' to convert 'Value' to Haskell values.
--   * 'toValue' to convert Haskell values to 'Value'.
--   * 'encodeJSON' to directly write Haskell value into JSON bytes.
--
-- For example,
--
-- > {-# LANGUAGE DeriveGeneric, DeriveAnyClass, DerivingStrategies #-}
-- >
-- > import GHC.Generics (Generic)
-- > import qualified Z.Data.Builder as Builder
-- > import qualified Z.Data.JSON as JSON
-- > import qualified Z.Data.Text as T
-- >
-- > data Person = Person {name :: T.Text, age :: Int}
-- >     deriving (Show, Generic)
-- >     deriving anyclass (JSON.JSON)
--
-- We can now encode & decode JSON like this:
--
-- >>> JSON.toValue (Person{ name="Alice", age=16 })
-- Object [("name",String "Alice"),("age",Number 16.0)]
-- >>> JSON.encode (Person{ name="Alice", age=16 })
-- [123,34,110,97,109,101,34,58,34,65,108,105,99,101,34,44,34,97,103,101,34,58,49,54,125]
-- >>> JSON.encodeText (Person{ name="Alice", age=16 })
-- "{\"age\":16,\"name\":\"Alice\"}"
-- >>> JSON.decodeText' "{\"age\":16,\"name\":\"Alice\"}" :: Either JSON.DecodeError Person
-- Right (Person {age = 16, name = "Alice"})
--
-- The 'GHC.Generics.Generic' based instances convert Haskell data with following rules:
--
--   * Constructors without payloads are encoded as JSON String, @data T = A | B@ are encoded as @\"A\"@ or @\"B\"@.
--   * Single constructor are ingored if there're payloads, @data T = T ...@,  @T@ is ingored:
--
--     * Records are encoded as JSON object. @data T = T{k1 :: .., k2 :: ..}@ are encoded as @{\"k1\":...,\"k2\":...}@.
--     * Plain product are encoded as JSON array. @data T = T t1 t2@ are encoded as "[x1,x2]".
--     * Single field plain product are encoded as it is, i.e. @data T = T t@ are encoded as \"x\" just like its payload.
--
--   * Multiple constructors are convert to single key JSON object if there're payloads:
--
--     * Records are encoded as JSON object like above. @data T = A | B {k1 :: .., k2 :: ..}@ are encoded as
--         @{\"B\":{\"k1\":...,\"k2\":...}}@ in @B .. ..@ case, or @\"A\"@ in @A@ case.
--     * Products inside a sum type are similar to above, wrapped by an outer single-key object layer marking which constructor used during data construction.
--
-- These rules apply to user defined ADTs, but some built-in instances have different behaviours, namely:
--
--   * @Maybe a@ are encoded as JSON @null@ in 'Nothing' case, or directly encoded to its payload in 'Just' case.
--   * @[a]@ are encoded to JSON array, @[Char]@ are encoded into JSON string.
--   * 'NonEmpty', 'Vector', 'PrimVector', 'HashSet', 'FlatSet', 'FlatIntSet' are also encoded to JSON array.
--   * 'Bytes' are encoded into JSON text using base64 encoding.
--   * 'HashMap', 'FlatMap', 'FlatIntMap' are encoded to JSON object.

-- $custom-settings
--
-- There're some modifying options if you providing a custom 'Settings', which
-- allow you to modify field name or constructor name, but please /DO NOT/
-- produce control characters during your modification, since we assume field
-- labels and constructor name won't contain them, thus we can save an extra
-- escaping pass. To use custom 'Settings' just write:
--
-- > data T = T {fooT :: Int, barT :: [Int]} deriving Generic
-- > instance JSON.JSON T where
-- >     -- You can omit following definition if you don't need to change settings
-- >     toValue = JSON.gToValue JSON.defaultSettings{ JSON.fieldFmt = JSON.snakeCase } . from
-- >     encodeJSON = JSON.gEncodeJSON JSON.defaultSettings{ JSON.fieldFmt = JSON.snakeCase } . from
--
-- >>> JSON.toValue (T 0 [1,2,3])
-- Object [("foo_t",Number 0.0),("bar_t",Array [Number 1.0,Number 2.0,Number 3.0])]
--
-- $manually-instance
--
-- You can write 'JSON' instances by hand if the 'Generic' based one doesn't suit you.
-- Here is an example similar to aeson's.
--
-- @
-- import qualified Z.Data.Text          as T
-- import qualified Z.Data.Vector        as V
-- import qualified Z.Data.Builder       as B
-- import qualified Z.Data.JSON          as JSON
-- import           Z.Data.JSON          ((.:), (.=), (.!), JSON(..))
--
-- data Person = Person { name :: T.Text , age  :: Int } deriving Show
--
-- instance JSON Person where
--     fromValue = JSON.withFlatMapR \"Person\" $ \\ v -> Person
--                     \<$\> v .: \"name\"
--                     \<*\> v .: \"age\"
--
--     toValue (Person n a) = JSON.object [\"name\" .= n, \"age\" .= a]
--
--     encodeJSON (Person n a) = JSON.object' $ (\"name\" .! n <> \"age\" .! a)
-- @
--
-- >>> toValue (Person "Joe" 12)
-- Object [("name",String "Joe"),("age",Number 12.0)]
-- >>> JSON.convert' @Person . JSON.Object $ V.pack [("name",JSON.String "Joe"),("age",JSON.Number 12.0)]
-- Right (Person {name = "Joe", age = 12})
-- >>> JSON.encodeText (Person "Joe" 12)
-- "{"name":"Joe","age":12}"
--
-- The 'Value' type is different from aeson's one in that we use @Vector (Text, Value)@ to represent JSON objects, thus
-- we can choose different strategies on key duplication, the lookup map type, etc. so instead of a single 'withObject',
-- we provide 'withHashMap', 'withHashMapR', 'withFlatMap' and 'withFlatMapR' which use different lookup map type, and different key order priority. Most of the time 'FlatMap' is faster than 'HashMap' since we only use the lookup map once, the cost of constructing a 'HashMap' is higher. If you want to directly work on key-values, 'withKeyValues' provide key-values vector access.
--
-- There're some useful tools to help write encoding code in "Z.Data.JSON.Builder" module, such as a JSON string escaping tool, etc.
--
-- If you don't particularly care for fast encoding, you can also use 'toValue' together with value builder, the overhead is usually very small.


-- | Snake casing a pascal cased constructor name or camel cased field name, words are always lower cased and separated by an
-- underscore.
snakeCase :: String -> T.Text
{-# INLINE snakeCase #-}
snakeCase = symbCase '_'

-- | Train casing a pascal cased constructor name or camel cased field name, words are always lower cased and separated by
-- a hyphen.
trainCase :: String -> T.Text
{-# INLINE trainCase #-}
trainCase = symbCase '-'

--------------------------------------------------------------------------------

symbCase :: Char -> String -> T.Text
{-# INLINABLE symbCase #-}
symbCase sym =  T.pack . go . applyFirst toLower
  where
    go []                       = []
    go (x:xs) | isUpper x = sym : toLower x : go xs
              | otherwise = x : go xs

    applyFirst _ []     = []
    applyFirst f (x:xs) = f x: xs

--------------------------------------------------------------------------------

instance JSON ExitCode where
    {-# INLINE fromValue #-}
    fromValue (String "ExitSuccess") = return ExitSuccess
    fromValue (Number x) =
        case toBoundedInteger x of
            Just i -> return (ExitFailure i)
            _      -> fail' . B.unsafeBuildText $ do
                "converting ExitCode failed, value is either floating or will cause over or underflow: "
                B.scientific x
    fromValue _ =  fail' "converting ExitCode failed, expected a string or number"

    {-# INLINE toValue #-}
    toValue ExitSuccess     = String "ExitSuccess"
    toValue (ExitFailure n) = Number (fromIntegral n)

    {-# INLINE encodeJSON #-}
    encodeJSON ExitSuccess     = "\"ExitSuccess\""
    encodeJSON (ExitFailure n) = B.int n

-- | Only round trip 'versionBranch' as JSON array.
instance JSON Version where
    {-# INLINE fromValue #-}
    fromValue v = makeVersion <$> fromValue v
    {-# INLINE toValue #-}
    toValue = toValue . versionBranch
    {-# INLINE encodeJSON #-}
    encodeJSON = encodeJSON . versionBranch

--------------------------------------------------------------------------------

-- | @YYYY-MM-DDTHH:MM:SS.SSSZ@
instance JSON UTCTime where
    {-# INLINE fromValue #-}
    fromValue = withText "UTCTime" $ \ t ->
        case P.parse' (P.utcTime <* P.endOfInput) (T.getUTF8Bytes t) of
            Left err -> fail' $ "could not parse date as UTCTime: " <> T.toText err
            Right r  -> return r
    {-# INLINE toValue #-}
    toValue t = String (B.unsafeBuildText (B.utcTime t))
    {-# INLINE encodeJSON #-}
    encodeJSON = B.quotes . B.utcTime

-- | @YYYY-MM-DDTHH:MM:SS.SSSZ@
instance JSON ZonedTime where
    {-# INLINE fromValue #-}
    fromValue = withText "ZonedTime" $ \ t ->
        case P.parse' (P.zonedTime <* P.endOfInput) (T.getUTF8Bytes t) of
            Left err -> fail' $ "could not parse date as ZonedTime: " <> T.toText err
            Right r  -> return r
    {-# INLINE toValue #-}
    toValue t = String (B.unsafeBuildText (B.zonedTime t))
    {-# INLINE encodeJSON #-}
    encodeJSON = B.quotes . B.zonedTime

-- | @YYYY-MM-DD@
instance JSON Day where
    {-# INLINE fromValue #-}
    fromValue = withText "Day" $ \ t ->
        case P.parse' (P.day <* P.endOfInput) (T.getUTF8Bytes t) of
            Left err -> fail' $ "could not parse date as Day: " <> T.toText err
            Right r  -> return r
    {-# INLINE toValue #-}
    toValue t = String (B.unsafeBuildText (B.day t))
    {-# INLINE encodeJSON #-}
    encodeJSON = B.quotes . B.day


-- | @YYYY-MM-DDTHH:MM:SS.SSSZ@
instance JSON LocalTime where
    {-# INLINE fromValue #-}
    fromValue = withText "LocalTime" $ \ t ->
        case P.parse' (P.localTime <* P.endOfInput) (T.getUTF8Bytes t) of
            Left err -> fail' $ "could not parse date as LocalTime: " <> T.toText err
            Right r  -> return r
    {-# INLINE toValue #-}
    toValue t = String (B.unsafeBuildText (B.localTime t))
    {-# INLINE encodeJSON #-}
    encodeJSON = B.quotes . B.localTime

-- | @HH:MM:SS.SSS@
instance JSON TimeOfDay where
    {-# INLINE fromValue #-}
    fromValue = withText "TimeOfDay" $ \ t ->
        case P.parse' (P.timeOfDay <* P.endOfInput) (T.getUTF8Bytes t) of
            Left err -> fail' $ "could not parse time as TimeOfDay: " <> T.toText err
            Right r  -> return r
    {-# INLINE toValue #-}
    toValue t = String (B.unsafeBuildText (B.timeOfDay t))
    {-# INLINE encodeJSON #-}
    encodeJSON = B.quotes . B.timeOfDay

-- | This instance includes a bounds check to prevent maliciously
-- large inputs to fill up the memory of the target system. You can
-- newtype 'NominalDiffTime' and provide your own instance using
-- 'withScientific' if you want to allow larger inputs.
instance JSON NominalDiffTime where
    {-# INLINE fromValue #-}
    fromValue = withBoundedScientific "NominalDiffTime" $ pure . realToFrac
    {-# INLINE toValue #-}
    toValue = Number . realToFrac
    {-# INLINE encodeJSON #-}
    encodeJSON = B.scientific' . realToFrac

-- | This instance includes a bounds check to prevent maliciously
-- large inputs to fill up the memory of the target system. You can
-- newtype 'DiffTime' and provide your own instance using
-- 'withScientific' if you want to allow larger inputs.
instance JSON DiffTime where
    {-# INLINE fromValue #-}
    fromValue = withBoundedScientific "DiffTime" $ pure . realToFrac
    {-# INLINE toValue #-}
    toValue = Number . realToFrac
    {-# INLINE encodeJSON #-}
    encodeJSON = B.scientific' . realToFrac

-- | @{"seconds": SSS, "nanoseconds": NNN}@.
instance JSON SystemTime where
    {-# INLINE fromValue #-}
    fromValue = withFlatMapR "SystemTime" $ \ v ->
        MkSystemTime <$> v .: "seconds" <*> v .: "nanoseconds"
    {-# INLINE toValue #-}
    toValue (MkSystemTime s ns) = object [ "seconds" .= s , "nanoseconds" .= ns ]
    {-# INLINE encodeJSON #-}
    encodeJSON (MkSystemTime s ns) = object' ("seconds" .! s <> "nanoseconds" .! ns)

instance JSON CalendarDiffTime where
    {-# INLINE fromValue #-}
    fromValue = withFlatMapR "CalendarDiffTime" $ \ v ->
        CalendarDiffTime <$> v .: "months" <*> v .: "time"
    {-# INLINE toValue #-}
    toValue (CalendarDiffTime m nt) = object [ "months" .= m , "time" .= nt ]
    {-# INLINE encodeJSON #-}
    encodeJSON (CalendarDiffTime m nt) = object' ("months" .! m <> "time" .! nt)

instance JSON CalendarDiffDays where
    {-# INLINE fromValue #-}
    fromValue = withFlatMapR "CalendarDiffDays" $ \ v ->
        CalendarDiffDays <$> v .: "months" <*> v .: "days"
    {-# INLINE toValue #-}
    toValue (CalendarDiffDays m d) = object ["months" .= m, "days" .= d]
    {-# INLINE encodeJSON #-}
    encodeJSON (CalendarDiffDays m d) = object' ("months" .! m <> "days" .! d)

instance JSON DayOfWeek where
    {-# INLINE fromValue #-}
    fromValue (String "Monday"   ) = pure Monday
    fromValue (String "Tuesday"  ) = pure Tuesday
    fromValue (String "Wednesday") = pure Wednesday
    fromValue (String "Thursday" ) = pure Thursday
    fromValue (String "Friday"   ) = pure Friday
    fromValue (String "Saturday" ) = pure Saturday
    fromValue (String "Sunday"   ) = pure Sunday
    fromValue (String _   )        = fail' "converting DayOfWeek failed, value should be one of weekdays"
    fromValue v                    = typeMismatch "DayOfWeek" "String" v

    {-# INLINE toValue #-}
    toValue Monday    = String "Monday"
    toValue Tuesday   = String "Tuesday"
    toValue Wednesday = String "Wednesday"
    toValue Thursday  = String "Thursday"
    toValue Friday    = String "Friday"
    toValue Saturday  = String "Saturday"
    toValue Sunday    = String "Sunday"

    {-# INLINE encodeJSON #-}
    encodeJSON Monday    = "\"Monday\""
    encodeJSON Tuesday   = "\"Tuesday\""
    encodeJSON Wednesday = "\"Wednesday\""
    encodeJSON Thursday  = "\"Thursday\""
    encodeJSON Friday    = "\"Friday\""
    encodeJSON Saturday  = "\"Saturday\""
    encodeJSON Sunday    = "\"Sunday\""

instance JSON UUID where
    {-# INLINE fromValue #-}
    fromValue = withText "UUID" $ \ t ->
        case P.parse' (P.uuid <* P.endOfInput) (T.getUTF8Bytes t) of
            Left err -> fail' $ "could not parse UUID: " <> T.toText err
            Right r  -> return r
    {-# INLINE toValue #-}
    toValue = String . B.unsafeBuildText . B.quotes . B.uuid
    {-# INLINE encodeJSON #-}
    encodeJSON = B.quotes . B.uuid

--------------------------------------------------------------------------------

deriving newtype instance JSON (f (g a)) => JSON (Compose f g a)
deriving newtype instance JSON a => JSON (Semigroup.Min a)
deriving newtype instance JSON a => JSON (Semigroup.Max a)
deriving newtype instance JSON a => JSON (Semigroup.First a)
deriving newtype instance JSON a => JSON (Semigroup.Last a)
deriving newtype instance JSON a => JSON (Semigroup.WrappedMonoid a)
deriving newtype instance JSON a => JSON (Semigroup.Dual a)
deriving newtype instance JSON a => JSON (Monoid.First a)
deriving newtype instance JSON a => JSON (Monoid.Last a)
deriving newtype instance JSON a => JSON (Identity a)
deriving newtype instance JSON a => JSON (Const a b)
deriving newtype instance JSON b => JSON (Tagged a b)

-- | Use 'Null' as @Proxy a@
instance JSON (Proxy a) where
    {-# INLINE fromValue #-}; fromValue = fromNull "Proxy" Proxy;
    {-# INLINE toValue #-}; toValue _ = Null;
    {-# INLINE encodeJSON #-}; encodeJSON _ = "null";

--------------------------------------------------------------------------------

deriving newtype instance JSON CChar
deriving newtype instance JSON CSChar
deriving newtype instance JSON CUChar
deriving newtype instance JSON CShort
deriving newtype instance JSON CUShort
deriving newtype instance JSON CInt
deriving newtype instance JSON CUInt
deriving newtype instance JSON CLong
deriving newtype instance JSON CULong
deriving newtype instance JSON CPtrdiff
deriving newtype instance JSON CSize
deriving newtype instance JSON CWchar
deriving newtype instance JSON CSigAtomic
deriving newtype instance JSON CLLong
deriving newtype instance JSON CULLong
deriving newtype instance JSON CBool
deriving newtype instance JSON CIntPtr
deriving newtype instance JSON CUIntPtr
deriving newtype instance JSON CIntMax
deriving newtype instance JSON CUIntMax
deriving newtype instance JSON CClock
deriving newtype instance JSON CTime
deriving newtype instance JSON CUSeconds
deriving newtype instance JSON CSUSeconds
deriving newtype instance JSON CFloat
deriving newtype instance JSON CDouble

--------------------------------------------------------------------------------

deriving anyclass instance (JSON (f a), JSON (g a), JSON a) => JSON (Sum f g a)
deriving anyclass instance (JSON a, JSON b) => JSON (Either a b)
deriving anyclass instance (JSON (f a), JSON (g a)) => JSON (Product f g a)

deriving anyclass instance (JSON a, JSON b) => JSON (a, b)
deriving anyclass instance (JSON a, JSON b, JSON c) => JSON (a, b, c)
deriving anyclass instance (JSON a, JSON b, JSON c, JSON d) => JSON (a, b, c, d)
deriving anyclass instance (JSON a, JSON b, JSON c, JSON d, JSON e) => JSON (a, b, c, d, e)
deriving anyclass instance (JSON a, JSON b, JSON c, JSON d, JSON e, JSON f) => JSON (a, b, c, d, e, f)
deriving anyclass instance (JSON a, JSON b, JSON c, JSON d, JSON e, JSON f, JSON g) => JSON (a, b, c, d, e, f, g)
